src/cyclefold/circuit.rs

```diff
@@ -0,0 +1,222 @@
+//! This module defines Cyclefold stuff
+
+use bellpepper_core::{boolean::AllocatedBit, ConstraintSystem, SynthesisError};
+
+use crate::{
+  gadgets::ecc::AllocatedPoint,
+  traits::{commitment::CommitmentTrait, Engine},
+  Commitment,
+};
+
+/// TODO: docs
+pub struct CyclefoldCircuitInputs<E: Engine> {
+  commit_1: Option<Commitment<E>>,
+  commit_2: Option<Commitment<E>>,
+  result: Option<Commitment<E>>,
+  scalar: Vec<Option<bool>>,
+}
+
+impl<E: Engine> CyclefoldCircuitInputs<E> {
+  /// TODO
+  pub fn new(
+    commit_1: Option<Commitment<E>>,
+    commit_2: Option<Commitment<E>>,
+    result: Option<Commitment<E>>,
+    scalar: Vec<Option<bool>>,
+  ) -> Self {
+    Self {
+      commit_1,
+      commit_2,
+      result,
+      scalar,
+    }
+  }
+}
+
+/// TODO: docs
+pub struct CyclefoldCircuit<E: Engine> {
+  inputs: Option<CyclefoldCircuitInputs<E>>,
+}
```

Comment:
What's the benefit of this indirection? Why not just implement `synthesize`, etc. for the 'inputs'?

Reply:
I get that you're wrapping in an `Option`, but this is already accomplished by the internal `Option`s in the contained struct. Moreover, the field is a `Vec` of `Option`s, and even when synthesizing shape, this should still be manifested in order to make the size of the inputs explicit, etc.

I maybe missing something, but I would probably reduce this to a single struct, perhaps with an initializer making it easy to create a (correctly-sized) empty version; and choose a name reflecting the result.